pkg/collector/corechecks/containers/containerd_test.go

```diff
@@ -0,0 +1,286 @@
+	"github.com/docker/docker/pkg/testutil/assert"
+	types2 "github.com/gogo/protobuf/types"
+	"github.com/stretchr/testify/mock"
+	"github.com/stretchr/testify/require"
+)
+
+type mockContainer struct {
+	containerd.Container
+	mockTask   func() (containerd.Task, error)
+	mockImage  func() (containerd.Image, error)
+	mockLabels func() (map[string]string, error)
+	mockInfo   func() (containers.Container, error)
+}
+
+// Task is from the containerd.Container interface
+func (cs *mockContainer) Task(context.Context, cio.Attach) (containerd.Task, error) {
+	return cs.mockTask()
+}
+
+// Image is from the containerd.Container interface
+func (cs *mockContainer) Image(context.Context) (containerd.Image, error) {
+	return cs.mockImage()
+}
+
+// Labels is from the containerd.Container interface
+func (cs *mockContainer) Labels(context.Context) (map[string]string, error) {
+	return cs.mockLabels()
+}
+
+// Info is from the containerd.Container interface
+func (cs *mockContainer) Info(context.Context) (containers.Container, error) {
+	return cs.mockInfo()
+}
+
+type mockTaskStruct struct {
+	containerd.Task
+	mockMectric func(ctx context.Context) (*types.Metric, error)
+}
+
+// Metrics is from the containerd.Task interface
+func (t *mockTaskStruct) Metrics(ctx context.Context) (*types.Metric, error) {
+	return t.mockMectric(ctx)
+}
+
+type mockImage struct {
+	imageName string
+	containerd.Image
+}
+
+// Name is from the Image interface
+func (i *mockImage) Name() string {
+	return i.imageName
+}
+
+// TestCollectTags checks the collectTags method
+func TestCollectTags(t *testing.T) {
+	img := &mockImage{}
+	tests := []struct {
+		name      string
+		labels    map[string]string
+		imageName string
+		runtime   string
+		expected  []string
+		err       error
+	}{
+		{
+			"all functioning",
+			map[string]string{"foo": "bar"},
+			"redis",
+			"containerd",
+			[]string{"runtime:containerd", "image:redis", "foo:bar"},
+			nil,
+		}, {
+			"missing labels",
+			map[string]string{},
+			"imagename",
+			"containerd",
+			[]string{"runtime:containerd", "image:imagename"},
+			nil,
+		},
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			cs := &mockContainer{
+				mockImage: func() (containerd.Image, error) {
+					img.imageName = test.imageName
+					return containerd.Image(img), nil
+				},
+				mockLabels: func() (map[string]string, error) {
+					return test.labels, nil
+				},
+				mockInfo: func() (containers.Container, error) {
+					ctn := containers.Container{
+						Runtime: containers.RuntimeInfo{
+							Name: test.runtime,
+						},
+					}
+					return ctn, nil
+				},
+			}
+			ctn := containerd.Container(cs)
+			list, err := collectTags(ctn, context.Background())
+			if err != nil {
+				require.Error(t, test.err, err)
+			}
+			sort.Strings(list)
+			sort.Strings(test.expected)
+			require.Equal(t, len(test.expected), len(list))
+			require.True(t, reflect.DeepEqual(test.expected, list))
```

Comment:
[`require.EqualValues`](https://godoc.org/github.com/stretchr/testify/require#EqualValues) should do what you want here, it's running a [DeepEqual](https://github.com/stretchr/testify/blob/c45a1382af909f3ed0abc3b8d8e59e6db48ed9d8/assert/assertions.go#L76-L94) under the hood. Also, if the objects are equal, the lengths will also be

Reply:
:til: